contracts/router/quoter/PoolQuoterV1.sol

```diff
@@ -0,0 +1,396 @@
+
+    /// @dev Checks whether `tokenIn -> tokenOut` is possible via the given Default Pool, given the
+    /// `actionMask` of available actions for the token.
+    /// Note: only checks DefaultPool-related actions: Swap/AddLiquidity/RemoveLiquidity.
+    function _isConnectedViaDefaultPool(
+        uint256 actionMask,
+        address pool,
+        address tokenIn,
+        address tokenOut
+    ) internal view returns (bool) {
+        // We don't check for paused pools here, as we only need to know if a connection exists.
+        (uint8 indexIn, uint8 indexOut) = _getTokenIndexes(pool, tokenIn, tokenOut);
+        // Check if Swap (tokenIn -> tokenOut) could fulfill tokenIn -> tokenOut request.
+        if (Action.Swap.isIncluded(actionMask) && indexIn != NOT_FOUND && indexOut != NOT_FOUND) {
+            return true;
+        }
+        address lpToken = _lpToken(pool);
+        // Check if AddLiquidity (tokenIn -> lpToken) could fulfill tokenIn -> tokenOut request.
+        if (Action.AddLiquidity.isIncluded(actionMask) && indexIn != NOT_FOUND && tokenOut == lpToken) {
+            return true;
+        }
+        // Check if RemoveLiquidity (lpToken -> tokenOut) could fulfill tokenIn -> tokenOut request.
+        if (Action.RemoveLiquidity.isIncluded(actionMask) && tokenIn == lpToken && indexOut != NOT_FOUND) {
+            return true;
+        }
+        return false;
+    }
+
+    /// @dev Checks whether `tokenIn -> tokenOut` is possible via the given Linked Pool, given the
+    /// `actionMask` of available actions for the token.
+    /// Note: only checks LinkedPool-related actions: Swap.
+    function _isConnectedViaLinkedPool(
+        uint256 actionMask,
+        address pool,
+        address tokenIn,
+        address tokenOut
+    ) internal view returns (bool) {
+        // Check if Swap (tokenIn -> tokenOut) could fulfill tokenIn -> tokenOut request.
+        if (Action.Swap.isIncluded(actionMask)) {
+            // Check if tokenIn and tokenOut are connected via the LinkedPool.
+            // We are converting ETH -> WETH here, as LinkedPool is unaware of ETH.
+            return ILinkedPool(pool).areConnectedTokens(_poolToken(tokenIn), _poolToken(tokenOut));
+        }
+        return false;
+    }
+
+    /// @dev Gets a quote for `tokenIn -> tokenOut` via the given Default Pool, given the `actionMask` of
+    /// available actions for the token.
+    /// If the action is possible, the `query` will be updated, should the output amount be better than
+    /// the current best quote.
+    /// Note: only checks pool-related actions: Swap/AddLiquidity/RemoveLiquidity.
+    /// Note: this is not supposed to be used with LinkedPool contracts, as a single token can appear
+    /// multiple times in the LinkedPool's token tree.
+    function _checkDefaultPoolQuote(
+        uint256 actionMask,
+        address pool,
+        address tokenIn,
+        address tokenOut,
```

Comment:
is `tokenOut` necessary as input param given `SwapQuery` also has `tokenOut` in the struct?

Reply:
`query` in `_check...Quote()` functions is the currently best found quote, which might have been populated yet, so `tokenOut` is passed separately.

Reply:
ah ok